dags/data_quality_checks_dag.py

```diff
@@ -0,0 +1,173 @@
+"""
+Imports local files to S3, then to CrateDB and checks several data quality properties
+
+Prerequisites
+-------------
+In CrateDB, set up tables for temporarily and permanently storing incoming data.
+See the file setup/smart_home_data.sql in this repository.
+"""
+import os
+import logging
+from airflow.utils.task_group import TaskGroup
+from airflow import DAG
+from airflow.utils.dates import datetime
+from airflow.decorators import task
+
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+from airflow.contrib.operators.slack_webhook_operator import SlackWebhookOperator
+from airflow.providers.common.sql.operators.sql import SQLColumnCheckOperator, SQLTableCheckOperator
+from airflow.providers.amazon.aws.transfers.local_to_s3 import LocalFilesystemToS3Operator
+from airflow.providers.amazon.aws.operators.s3_copy_object import S3CopyObjectOperator
+from airflow.providers.amazon.aws.operators.s3_delete_objects import S3DeleteObjectsOperator
+
+from include.data_checks import COL_CHECKS, TABLE_CHECKS
+
+
+S3_BUCKET = os.environ.get("S3_BUCKET")
+ACCESS_KEY_ID = os.environ.get("SECRET_ACCESS_KEY")
+SECRET_ACCESS_KEY = os.environ.get("SECRET_ACCESS_KEY")
+FILE_DIR = os.environ.get("FILE_DIR")
+TEMP_TABLE = os.environ.get("TEMP_TABLE")
+TABLE = os.environ.get("TABLE")
+
+
+def slack_failure_notification(context):
+    task_id = context.get('task_instance').task_id
+    dag_id = context.get('task_instance').dag_id
+    exec_date = context.get('execution_date')
+    log_url = context.get('task_instance').log_url
+    slack_msg = f"""
+            :red_circle: Task Failed. 
+            *Task*: {task_id}  
+            *Dag*: {dag_id} 
+            *Execution Time*: {exec_date}  
+            *Log Url*: {log_url} 
+            """
+    failed_alert = SlackWebhookOperator(
+        task_id='slack_notification',
+        http_conn_id='slack_webhook',
+        message=slack_msg)
+    return failed_alert.execute(context=context)
+
+
+@task
+def get_files_from_s3(bucket, prefix_value):
+    s3_hook = S3Hook()
+    paths = s3_hook.list_keys(bucket_name=bucket, prefix=prefix_value)
+    return paths
+
+
+@task
+def get_import_statements(bucket, prefix_value):
+    s3_hook = S3Hook()
+    file_paths = s3_hook.list_keys(bucket_name=bucket, prefix=prefix_value)
+    statements = []
+    for path in file_paths:
+        sql = f"""
+                COPY {TEMP_TABLE} FROM 's3://{ACCESS_KEY_ID}:{SECRET_ACCESS_KEY}@{S3_BUCKET}/{path}' WITH (format='csv');
+              """
+        logging.info(sql)
+        statements.append(sql)
+    return statements
+
+
+with DAG(
+    "data_quality_checks",
+    description="DAG for checking quality of home metering data.",
+    start_date=datetime(2021, 1, 1),
+    schedule_interval=None,
+    catchup=False,
+    on_failure_callback=slack_failure_notification
```

Comment:
Do we want the failure callback to run if *any* task fails, or only particular ones? If it's all of them, then I think this should be under the `default_args` parameter to pass to each task.

Reply:
Done!